From 214fdf472f3b797f8b88f9c57488fa98cbb0515e Mon Sep 17 00:00:00 2001
From: Yusuke Sato <yusukes@chromium.org>
Date: Tue, 19 Oct 2010 20:30:09 +0900
Subject: [PATCH 09/10] Fix race condition in bus_ibus_impl_create_engine()

If the bus_ibus_impl_create_engine() function is called right after an ibus_component_start() call, the function might fail getting a factory object.

To avoid the problem, we should use the busy-wait logic even when ibus_component_is_running() returns true.

BUG=http://crosbug.com/7244
TEST=see the bug (comment #4,7,9)

Review URL: http://codereview.appspot.com/2562041
---
 bus/ibusimpl.c |   38 ++++++++++++++++++++++----------------
 1 files changed, 22 insertions(+), 16 deletions(-)

diff --git a/bus/ibusimpl.c b/bus/ibusimpl.c
index 620b56d..d84ba49 100644
--- a/bus/ibusimpl.c
+++ b/bus/ibusimpl.c
@@ -819,6 +819,27 @@ _ibus_get_address (BusIBusImpl     *ibus,
     return reply;
 }
 
+static BusFactoryProxy *
+_get_factory_proxy(IBusEngineDesc *engine_desc)
+{
+    BusFactoryProxy *factory;
+    GTimeVal t1, t2;
+    g_get_current_time (&t1);
+    while (1) {
+        if (g_main_context_pending (NULL)) {
+            g_main_context_iteration (NULL, FALSE);
+            factory = bus_factory_proxy_get_from_engine (engine_desc);
+            if (factory != NULL) {
+                return factory;
+            }
+        }
+        g_get_current_time (&t2);
+        if (t2.tv_sec - t1.tv_sec >= 5)
+            break;
+    }
+    return NULL;
+}
+
 static BusEngineProxy *
 bus_ibus_impl_create_engine (IBusEngineDesc *engine_desc)
 {
@@ -836,23 +857,8 @@ bus_ibus_impl_create_engine (IBusEngineDesc *engine_desc)
 
         if (!ibus_component_is_running (comp)) {
             ibus_component_start (comp, g_verbose);
-            g_get_current_time (&t1);
-            while (1) {
-                if (g_main_context_pending (NULL)) {
-                    g_main_context_iteration (NULL, FALSE);
-                    factory = bus_factory_proxy_get_from_engine (engine_desc);
-                    if (factory != NULL) {
-                        break;
-                    }
-                }
-                g_get_current_time (&t2);
-                if (t2.tv_sec - t1.tv_sec >= 5)
-                    break;
-            }
-        }
-        else {
-            factory = bus_factory_proxy_get_from_engine (engine_desc);
         }
+        factory = _get_factory_proxy (engine_desc);
     }
 
     if (factory == NULL) {
-- 
1.7.1

